File: c#list/CList.cs
using System;

namespace ConsoleApp2
{
    class CList<T>
    {
        public T value;
        public int size
        {
            set
            {
                int i = 0;
                ListElement<T> Head = head;
                while (Head != null)
                {
                    i++;
                    Head = Head.next;
                }
                size = i;
            }
            get
            {
                int i = 0;
                ListElement<T> Head = head;
                while (Head != null)
                {
                    i++;
                    Head = Head.next;
                }
                return i;
            }
        }
        private ListElement<T> head;
        private ListElement<T> prev;
        private ListElement<T> list;
        private static ListElement<T> delEl;

        public void addElement(T number)
        {
            ListElement<T> l = new ListElement<T>(number);
            if (list != null)
            {
                GetLast().next = l;
                GetLast().prev = prev;
                prev = GetLast();
                return;
            }
            list = l;
            head = list;
            prev = list;
        }

        public void PushFront(T number)
        {
            ListElement<T> listE = new ListElement<T>(number);
            listE.next = head;
            head.prev = listE;
            head = listE;
        }

        public ListElement<T> PopFront()
        {
            GetLast().prev.next = null;
            delEl = GetLast();
            return delEl;
        }

        public ListElement<T> PopBack()
        {
            delEl = head;
            head.next.prev = null;
            head = head.next;
            return delEl;
        }

        public ListElement<T> Remove(int index)
        {
            if (index > size || index < 0)
            {
                throw new Exception("index > size or index < 0");
            }
            ListElement<T> current = head;
            ListElement<T> currentT = head;
            int i = 0;
            for (int j = 0; j < size; j++)
            {
                if (j == index)
                {
                    delEl = currentT;
                }
                currentT = currentT.next;
            }
            if (index != 0 && index <= size - 2)
            {
                for (int j = 0; j < size; j++)
                {
                    if (j == index)
                    {
                        current.prev.next = current.next;
                        current.next.prev = current.prev.next;
                    }
                    current = current.next;
                }
            }
            else if (index == 0 && index != size - 1)
            {
                current.next.prev = null;
                head = current.next;
            }
            else if (index != 0 && index >= size - 1)
            {
                if (GetLast() != null && GetLast().prev != null)
                {
                    GetLast().prev.next = null;
                }
                else
                {
                    GetLast().next = null;
                }
            }
            return delEl;
        }

        public void InsertBefore(int index, T number)
        {
            if (index > size || index < 0)
            {
                throw new Exception("index > size or index < 0");
            }
            ListElement<T> current = head;
            int i = 0;
            ListElement<T> e = new ListElement<T>(number);
            while (current.next != null && index <= size && index != 0)
            {
                if (i == index-1)
                {
                    e.prev = current;
                    e.next = current.next;
                    current.next = e;
                }
                current = current.next;
                i++;
            }
            if(index == 0)
            {
                current.prev = e;
                e.next = current;
                head = e;
            }
            if (current.next == null && index == size)
            {
               current.next = e;
            }

        }
        public void InsertAfter(int index, T number)
        {
            if (index > size - 1 || index < 0)
            {
                throw new Exception("index > size or index < 0");
            }
            ListElement<T> current = head;
            int i = 0;
            ListElement<T> e = new ListElement<T>(number);
            while (current.next != null && index != size)
            {
                if (i == index)
                {
                    e.prev = current;
                    e.next = current.next;
                    current.next = e;
                }
                current = current.next;
                i++;
            }
            if(current.next == null && index == size-1)
            {
                current.next = e;
                e.prev = current;
            }

        }

        public void InsertBefore(int index, ListElement<T> e)
        {
            ListElement<T> current = head;
            ListElement<T> TempCurrent;
            int i = 0;
            while (current.next != null)
            {
                if (i == index)
                {
                    e.prev = current;
                    e.next = current.next;
                    current.next = e;
                }
                current = current.next;
                i++;
            }
        }

        public ListElement<T> GetLast()
        {
            ListElement<T> current = list;
            while (current.next != null)
            {
                current = current.next;
            }
            return current;

        }

        public void Print()
        {
            ListElement<T> Head = head;
            while (Head != null)
            {
                Console.WriteLine(Head.value);
                Head = Head.next;
            }
            Console.WriteLine();
        }
    }
}

File: test/CList.java
package test;

import lombok.Data;
import lombok.Getter;
import lombok.Setter;
import lombok.var;

@Getter
@Setter
public class CList<T> {

	private int size;
	private ListElement<T> head;
	private ListElement<T> last;
	private ListElement<T> prev;

	public void add(T element) {
		sizeIncrease();
		ListElement<T> l = new ListElement<T>();
		l.setValue(element);
		if (last != null) {
			getLast().setNext(l);
			getLast().setPrev(prev);
			prev = getLast();
			return;
		}
		last = l;
		head = last;
		prev = last;
	}

	public void sizeIncrease() {
		size++;
	}

	public void sizeDecrease() {
		size--;
	}

	public void remove(int index) {
//		ListElement<T> element = getElement(index);
//		if (element == null) {
//			throw new IndexOutOfBoundsException();
//		}
//		if (element.getPrev() != null)
//		{
//			element.getPrev().setNext(element.getNext());
//		}
//		else
//		{
//			head = element.getNext();
//		}
//
//		if (element.getNext() != null)
//		{
//			element.getNext().setPrev(element.getPrev());
//		} else
//		{
//			last = element.getPrev();
//		}
//
//		element.setNext(null);
//		element.setPrev(null);
//		
//		sizeDecrease();
	
	}

	public void print() {
		ListElement<T> current = head;
		while (current.getNext() != null) {
			System.out.println(current.getValue());
			current = current.getNext();
		}
		if (current.getNext() == null) {
			System.out.println(current.getValue());
		}
	}

	public void printBack() {
		ListElement<T> last = getLast();
		while (last.getPrev() != null) {
			System.out.println(last.getValue());
			last = last.getPrev();
		}
		if (last.getPrev() == null) {
			System.out.println(last.getValue());
		}
	}

	public ListElement<T> popFront() {
		sizeDecrease();
		try {
			return getLast();
		} finally {
			getLast().getPrev().setNext(null);
			;
		}
	}

	public ListElement<T> popBack() {
		sizeDecrease();
		try {
			return head;
		} finally {
			head.getNext().setPrev(null);
			head = head.getNext();
		}
	}

	public ListElement<T> getElement(int index) {
		if (index >= size) {
			throw new IndexOutOfBoundsException();
		}
		ListElement<T> current = head;
		for (int i = 0; i < getSize(); i++) {
			if (i == index) {
				return current;
			}
			current = current.getNext();
		}

		return current;

	}

	public ListElement<T> getLast() {
		ListElement<T> current = last;
		while (current.getNext() != null) {
			current = current.getNext();
		}
		return current;

	}

}
